virtualpdu/libvirt_device_provider.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2016 Internap
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import libvirt
+from oslo_config import cfg
+
+from virtualpdu import device_provider
+
+opts = [
+    cfg.StrOpt('uri',
+               help='Libvirt URI'),
+]
+
+CONF = cfg.CONF
+CONF.register_opts(opts, group='libvirt')
+
+DOMAIN_NOT_FOUND_MESSAGES = ['virDomainLookupByName() failed',
+                             'Domain not found']
+
+
+class LibvirtDeviceProvider(device_provider.DeviceProvider):
+    def power_on(self, name):
+        with ConnectedLibvirt() as connection:
+            domain = safe_lookup_by_name(connection, name)
+            try:
+                domain.create()
+            except libvirt.libvirtError as e:
+                if 'is already running' not in str(e):
+                    raise
+
+    def power_off(self, name):
+        with ConnectedLibvirt() as connection:
+            domain = safe_lookup_by_name(connection, name)
+            domain.destroy()
+
+
+def safe_lookup_by_name(connection, name):
+    try:
+        domain = connection.lookupByName(name)
+    except libvirt.libvirtError as e:
+        if any([m for m in DOMAIN_NOT_FOUND_MESSAGES if m in str(e)]):
+            raise LibvirtDomainNotFound()
+    else:
+        return domain
+
+
+class ConnectedLibvirt(object):
+    def __enter__(self):
+        self.connection = libvirt.open(CONF.libvirt.uri)
```

Comment:
I was passing the config around before (provider had configs which it passed to the context) but it was bloating the code. My tests have the same number of lines and this one is actually easier to read (we read the config where we need it).

Is there any advantage other than that?